pkg/arvo/lib/lv.hoon

```diff
@@ -0,0 +1,546 @@
+  ++  inner
+    ~/  %inner
+    |=  [u=@lvs v=@lvs]  ^-  @rs
+    ~_  leaf+"lagoon-fail"
+    =/  n  `@ud`(length u)
+    =/  count  0
+    =/  sum  .0
+    |-  ^-  @rs
+      ?:  =(count n)  `@rs`sum
+    $(count +(count), sum (add:rs sum (mul:rs (get u +(count)) (get v +(count)))))
+  ++  outer  !!
+  --
+::
+::::  Vector type in double-precision floating-point @rd
+  ::
+++  lvd
+  ^|
+  ~%  %lvd  ..is  ~
+  |_  r/$?($n $u $d $z)   :: round nearest, round up, round down, round to zero
+  ::
+  ::  Manipulatord
+  ::
+  ::    Zeroes
+  ++  zeros
+    ~/  %zeros
+    |=  n=@ud  ^-  @lvd
+    ~_  leaf+"lagoon-fail"
+    :: There is a vanishingly small chance of getting a NaN in a vector which
+    :: collides with this, but it's like getting a hash table collision, very
+    :: very unlikely.  Also the Urbit NaN is hard-coded, `DOUBNAN`.
+    =/  nan  0xffff.ffff.ffff.ffff
+    `@lvd`(not 6 +(n) (rap 6 ~[(fil 6 n nan) (not 6 1 n)]))
+  ::
+  ::    Ones
+  ++  ones
+    ~/  %ones
+    |=  n=@ud  ^-  @lvd
+    ~_  leaf+"lagoon-fail"
+    =/  one  (bit:rd [%f s=%.y e=-0 a=1])
+    =/  not-one  (not 6 1 one)
+    `@lvd`(not 6 +(n) (rap 6 ~[(fil 6 n not-one) (not 6 1 n)]))
+  ::
+  ::    Length of vector
+  ++  length
+    |=  u=@lvd  ^-  @ud
+    ~_  leaf+"lagoon-fail"
+    (snag 0 (flop (rip 6 u)))
+  ::
+  ::    Produce a vector from `(list @rd)`
+  ++  not-rd
+    |=  a=@rd
+    ~_  leaf+"lagoon-fail"
+    (not 6 1 a)
+  ++  make
+    |=  [u=(list @rd)]  ^-  @lvd
+    ~_  leaf+"lagoon-fail"
+    =/  n  `@ud`(lent u)
+    =/  not-u  (turn u not-rd)
+    `@lvd`(not 6 +(n) (rep 6 (snoc not-u (not 6 1 n))))
+  ++  unmake
+    |=  [u=@lvd]  ^-  (list @rd)
+    ~_  leaf+"lagoon-fail"
+    %-  flop
+    =/  n  `@ud`(length u)
+    ::  Leading and trailing zeros get cut off (because we flop or swp):
+    ::  we have to pin values on both ends to compensate, then trim.
+    =/  nan  0xffff.ffff.ffff.ffff
+    =/  ut  (swp 6 (cat 6 nan (cat 6 u nan)))
+    (oust [n 1] (oust [0 2] (rip 6 ut)))
+  ::
+  ::    Get the value at an index, using mathematical indices 1..n.
+  ++  get
+    ~/  %get
+    |=  [u=@lvd i=@ud]  ^-  @rd
+    ~_  leaf+"lagoon-fail"
+    =/  n  `@ud`(length u)
+    (snag (dec i) (unmake u))
+  ::
+  ::    Pretty-print the contents of the vector.
+  ++  pprint  !!
+  ::
+  ::    Set the value of an element within a vector, using math indices 1..n.
+  ++  set
+    ~/  %set
+    |=  [u=@lvd i=@ud x=@rd]  ^-  @lvd
+    ~_  leaf+"lagoon-fail"
+    (make (weld (scag (dec i) (unmake u)) [x (slag i (unmake u))]))
+  ::
+  ::    Return argument of maximum value in array, 1-indexed
+  ++  argmax
+    |=  [u=@lvd]  ^-  @ud
+    ~_  leaf+"lagoon-fail"
+    =/  n  `@ud`(length u)
+    =/  count  0
+    =/  max-val  `@rs`0xfeff.ffff
+    =/  max-arg  0
+    |-  ^-  @ud
+      ?:  =(count n)  +(max-arg)
+      =/  cur-val  `@rs`(get u +(count))
+      ?:  (gth:rs cur-val max-val)  $(count +(count), max-arg count)
+    $(count +(count), max-arg max-arg)
```

Comment:
No `@lvs` variant? Also, I'd maybe go for initializing max from the first candidate
```
=/  max  n
|-  ^+  max
?~  n  max
=?  max  (gte:rs (get u n) (get u max))  n
$(n (dec n))
```
Also also, should `.nan` count as min, max, or error here? (I think in current formulation it's max)

Reply:
Also mixed @rd and @rs... yeesh.

Reply:
```hoon
  ::
  ::    Return larger of two single-precision floats.
  ++  max-rs
    |=  [a=@rs b=@rs]  ^-  @rs
    ?:  (gth:rs a b)  a  b
  ::
  ::    Find maximum value in array.
  ++  max
    |=  [u=@lvs]  ^-  @rs
    ~_  leaf+"lagoon-fail"
    `@rs`(reel (unmake u) max-rs)
  ::
  ::    Return index of maximum value in array, 1-indexed
  ::      array.index(max(array)) is simple and fast
  ++  argmax
    |=  [u=@lvs]  ^-  @ud
    ~_  leaf+"lagoon-fail"
    (find ~[(max u)] u)
```